### DISCUSSION TITLE
How next-auth connects to mongodb? Please help find this in the sources

### DISCUSSION THREAD

**Discussion by YuriGor** [nextauthjs/next-auth#1344] [Help]
Hi all, I tried to find how does next-auth connects to mongodb server using provided connection string.
I need it because I want to do the same in my app.
I found that my nextjs app connects to atlas fine when it's running using 'next' command.
But when I try to run it using 'vercel dev' locally, or when I deploy it to vercel - mongo throws "user is not allowed to do action" error.
Most interesting thing - next-auth is still working fine in all the cases, so I want to replicate next-auth's way of connecting to mongo in my own code.
Probably all the connection job is done by typeorm lib, so I need to look there?

**Comment by brunocrosier** [comment]
Not sure if I understood exactly what you mean but there are two options, from what I can tell:

1.  Copy and paste your mongodb connection string into `.env.local` [like this](https://github.com/nextauthjs/next-auth-example/blob/main/.env.local.example#L28) 
2. In your `[...nextauth].js` file you can pass a database object in the `options` [like this](https://next-auth.js.org/configuration/databases#how-to-use-a-database)

**Reply by YuriGor** [reply]
next-auth was already working, I was wondering why my app doesn't)

**Comment by YuriGor** [comment]
Ok found, in vercel env var db name had trailing new line

**Reply by iaincollins** [reply]
@YuriGor Thank you for following up!

Extra characters in environment variables (or characters that are stripped or not escaped properly) are probably the biggest cause of "works locally but not in production" type problems folks run into.

This is useful feedback as helps us understand common pitfalls and how we can help -  e.g. with console warnings or by automatically trimming parameters that might contain invalid / problematic characters.

We do already with `NEXTAUTH_URL` as a lot of folks were tripping up with it and I think there is scope to expand on this for options like database connection strings and client / secret values.

Part of the reason it's such a common problem is the tooling around how environment variables are configured and viewing them once configured and viewing logs on hosting platforms. I'm thinking of introducing some sort of debug page which folks can enable to display current configuration in production, and test/debug things like sending email.

**Reply by YuriGor** [reply]
It was just me, who started development of new project from the next-auth example app, because for me authorization was the most painful part (and thank you guys a lot for taking this pain away from me), but actually it's a question for vercel team and dashboard UI, not for next-auth team.

Maybe if I could read next-auth docs carefully, I would discover the existence of 'vercel env pull' command mentioned in this section:
https://next-auth.js.org/getting-started/example#deploying-to-production
and I could use it instead of copying all those env variables manually on vercel dashboard.
